### ISSUE TITLE
F1 score of 0. IN vs BN

### ISSUE THREAD

**Issue by ystan98** [SunnyHaze/IML-ViT#51]
Any suggestions for a single GPU that can only accommodate up to batch size of 2? Note that I'm using the json_dataset method, if it matters.

Most of my experiments resulted in 0 f1 score (at 20epochs) despite trying different random states. The only being the params below:

  --edge_lambda 40 \
  --predict_head_norm "IN" \

Would BN be better?

I'm getting F1 score of 0 even when my test and train data is the same.

Additionally, could you share CASIA1.0 as well?

**Comment by SunnyHaze** [comment]
hi, thanks for your attention to our work.

could you please provide your dataset for training? and other detailed hyperparameters in the shell script for us to locate the issue?

For CASIAv1 dataset, you can reference here https://github.com/namtpham/casia1groundtruth

Note that clone this repo to obtain the mask, while the original image is provided in the Google Drive in the readme file.

**Comment by ystan98** [comment]
Thanks for your reply.

I got the CASIA2.0 dataset from this repo

https://github.com/SunnyHaze/IML-Dataset-Corrections

Then, I used this to prepare
```

import os
import json

base_path = "train_data/casia"  # Adjust to your absolute path
tp_dir = os.path.join(base_path, "Tp")
gt_dir = os.path.join(base_path, "Gt")
au_dir = os.path.join(base_path, "Au")

pairs = []

for fname in sorted(os.listdir(tp_dir)):
    if fname.endswith(".jpg"):
        img_path = os.path.join(tp_dir, fname)
        gt_name = fname.replace(".jpg", "_gt.png")
        gt_path = os.path.join(gt_dir, gt_name)

        if os.path.exists(gt_path):
            pairs.append([img_path, gt_path])
        else:
            print(f" Warning: Ground truth not found for {fname}")

for fname in sorted(os.listdir(au_dir)):
    if fname.endswith(".jpg"):
        img_path = os.path.join(au_dir, fname)
        pairs.append([img_path, "Negative"])

with open("casia_train.json", "w") as f:
    json.dump(pairs, f, indent=2)

print(f" Saved {len(pairs)} samples to casia_train.json")
```

**Comment by SunnyHaze** [comment]
Hi, Everything is correct with your data processing.

Overall, I have two suggestions for you:
1. You may increase the number of `--accum_iter` to 32 to meet our setting when you only have one GPU. (4*GPU, batchsize=2, accum_iter=8)
2. Try training only with manipulated images at first.

**Comment by ystan98** [comment]
1) I will try it now
2) I tried this, managed to see F1>0. 

Is it advisable to continue going with #2? My understanding is including authentic images leads to better generalisation.

**Comment by SunnyHaze** [comment]
Including more authentic images can help the model to reduce the FALSE POSITIVE alert.

However, as you can see, the authentic images are images with all pixels==0, which may lead to the model trying to predict every image as black when including too many authentic images. Thus, I believe there should be a rate or balance the proportion between authentic and manipulated images.

**Comment by ystan98** [comment]
Thank you! @SunnyHaze 

I will include more tampered images together with CASIA2.0 perhaps.
Have your team tried modulating loss functions to account for the imbalanced data? (tampered vs authentic)

**Comment by SunnyHaze** [comment]
I believe `--edge_lambda 40` is part of an effort to solve the imbalance issue. Since it give larger weights to the edge area (have white pixels)

Anyway, the padding strategy in IML-ViT is another factor that exacerbates this issue. However, our team didn't make much effort on design a loss to further balance these issues. Since current --edge_lambda is efficient to make the model have F1 > 0 on 4 GPU experiments.

**Comment by ystan98** [comment]
Thanks! May I also ask if a larger batch_size (e.g: 4) is preferred (do-able if we decrease input dimensions from 1024 to 512), or is higher input dimensions preferred (batch_size needs to be smaller e.g: 2).

**Comment by SunnyHaze** [comment]
Hi,
From my experience, a larger batch size would give better performance. However, as high resolution is also important in IML tasks, I think applying a sliding window strategy with a resolution of 512x512 would be a good choice to gain all the profit simultaneously.

**Comment by ystan98** [comment]
Hi, thanks!

I'll try to reduce from 1024 to 512, and increase batch_size.

**Comment by SunnyHaze** [comment]
if you are trying apply 512, you may use SparseViT instead of IML-ViT.

https://github.com/scu-zjz/SparseViT

This one is more memory efficient. And can be called from [IMDL-BenCo](https://github.com/scu-zjz/IMDLBenCo) easily.

**Comment by ystan98** [comment]
Thanks, let me check up on it!

How's the comparison of the performances between sparseViT and IML-Vit respectively?

**Comment by SunnyHaze** [comment]
IMDLBenCo is a training framework. It contains both the implementation of IML-ViT and SparseViT for IMDL tasks. Further, IMDLBenCo also contains the implementation of eight other SoTA models as well.

For the performance, I think sparseViT works better on a large amount of data. If you are working with 1k~9k images, IML-ViT would be a better choice.

**Comment by ystan98** [comment]
Thanks for the explanation! I'll read the papers :)

**Comment by SunnyHaze** [comment]
For more details, you can refer to the [documentation of IMDL-BenCO](https://scu-zjz.github.io/IMDLBenCo-doc/)

**Comment by ystan98** [comment]
Any advice? Around 86th epoch this happened. My f1 before was quite good. Now became 0.



11:26:35.199785] Epoch: [86]  [3000/7072]  eta: 0:18:27  lr: 0.000063  predict_loss: 0.1042 (0.0969)  edge_loss: 0.0955 (0.0769)  time: 0.2705  data: 0.0202  max mem: 28558
[11:27:29.626591] Epoch: [86]  [3200/7072]  eta: 0:17:32  lr: 0.000063  predict_loss: 0.0597 (0.0978)  edge_loss: 0.0440 (0.0776)  time: 0.2726  data: 0.0214  max mem: 28558
[11:28:24.017015] Epoch: [86]  [3400/7072]  eta: 0:16:38  lr: 0.000062  predict_loss: 0.0549 (0.0974)  edge_loss: 0.0353 (0.0773)  time: 0.2736  data: 0.0223  max mem: 28558
[11:29:18.356439] Epoch: [86]  [3600/7072]  eta: 0:15:44  lr: 0.000062  predict_loss: 0.0876 (0.0975)  edge_loss: 0.0559 (0.0774)  time: 0.2719  data: 0.0206  max mem: 28558
[11:30:12.541319] Epoch: [86]  [3800/7072]  eta: 0:14:49  lr: 0.000062  predict_loss: 0.0536 (0.0974)  edge_loss: 0.0473 (0.0773)  time: 0.2701  data: 0.0195  max mem: 28558
[11:31:06.860552] Epoch: [86]  [4000/7072]  eta: 0:13:55  lr: 0.000062  predict_loss: 0.0731 (0.0971)  edge_loss: 0.0490 (0.0771)  time: 0.2733  data: 0.0221  max mem: 28558
[11:32:01.008758] Epoch: [86]  [4200/7072]  eta: 0:13:00  lr: 0.000062  predict_loss: 0.0835 (0.0968)  edge_loss: 0.0588 (0.0770)  time: 0.2713  data: 0.0200  max mem: 28558
[11:32:55.535483] Epoch: [86]  [4400/7072]  eta: 0:12:06  lr: 0.000062  predict_loss: nan (nan)  edge_loss: nan (nan)  time: 0.2792  data: 0.0241  max mem: 28558
[11:33:49.581011] Epoch: [86]  [4600/7072]  eta: 0:11:11  lr: 0.000062  predict_loss: nan (nan)  edge_loss: nan (nan)  time: 0.2677  data: 0.0181  max mem: 28558
[11:34:43.701199] Epoch: [86]  [4800/7072]  eta: 0:10:17  lr: 0.000062  predict_loss: nan (nan)  edge_loss: nan (nan)  time: 0.2687  data: 0.0192  max mem: 28558
[11:35:37.809675] Epoch: [86]  [5000/7072]  eta: 0:09:22  lr: 0.000062  predict_loss: nan (nan)  edge_loss: nan (nan)  time: 0.2682  data: 0.0186  max mem: 28558

**Comment by SunnyHaze** [comment]
From my point of view, here are some possible advice:
1. Prevent too much black area mask in the dataset. Since most of the pixels are black, the gradient will become extremely small and cause a nan issue.
2. Consider using the `gradient clipping method` to prevent this issue. Note that this may also need to revise something related with AMP and gradient scaler. I am not quite confident with it now, you could check the documentation or ask GPT for guidance.
